scripts/state_space_models_lib.py

```diff
@@ -0,0 +1,316 @@
+# Library of state space models using Deepmind's distrax framework
+# See: https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/LinearGaussianStateSpaceModel
+#      https://github.com/deepmind/distrax/blob/master/distrax/_src/distributions/distribution.py
+# for implementation details and specifications
+
+# Author: Gerardo Durán-Martín (@gerdm)
+
+# !pip install -q distrax
+
+import chex
+import jax
+import distrax
+import operator
+import functools
+import jax.numpy as jnp
+import collections.abc
+import numpy as np
+from jax import random
+from tensorflow_probability.substrates import jax as tfp
+from typing import Optional, Tuple, Union, Sequence
+from distrax._src.utils import jittable
+
+tfd = tfp.distributions
+Array = chex.Array
+Numeric = chex.Numeric
+PRNGKey = chex.PRNGKey
+IntLike = Union[int, np.int16, np.int32, np.int64]
+
+
+def MultivariateNormal(loc, covariance_matrix):
+    return distrax.as_distribution(tfd.MultivariateNormalFullCovariance(loc, covariance_matrix))
+
+
+def convert_seed_and_sample_shape(
+    seed: Union[IntLike, PRNGKey],
+    sample_shape: Union[IntLike, Sequence[IntLike]]) -> Tuple[PRNGKey, Tuple[int, ...]]:
+    """Shared functionality to ensure that seeds and shapes are the right type."""
+    if not isinstance(sample_shape, collections.abc.Sequence):
+        sample_shape = (sample_shape,)
+        sample_shape = tuple(map(int, sample_shape))
+
+    if isinstance(seed, IntLike.__args__):
+        rng = jax.random.PRNGKey(seed)
+    else:  # key is of type PRNGKey
+        rng = seed
+
+    return rng, sample_shape
+
+
+class LinearGaussianStateSpaceModel(jittable.Jittable):
+    def __init__(self, num_timesteps, transition_matrix, transition_noise,
+                 observation_matrix, observation_noise, initial_state_prior):
+        self._num_timesteps = num_timesteps
+        self._transition_matrix = transition_matrix
+        self._transition_noise = transition_noise
+        self._observation_matrix = observation_matrix
+        self._observation_noise = observation_noise
+        self._initial_state_prior = initial_state_prior
+        self.observation_size, self.state_size = observation_matrix.shape
+
+    @property
+    def num_timesteps(self) -> Array:
+        return self._num_timesteps
+
+    @property
```

Comment:
why do you need all these property methods? why not access the fields directly?

Reply:
To have immutable (ish) properties. It is a good convention  (see [property function](https://docs.python.org/3/library/functions.html#property))

This raises an error if the user writes 

```python
model = LinearGaussianStateSpaceModel(...)
model.transition_matrix = jnp.zeros((2, 2)) # This fails
```